server.go

```diff
@@ -299,6 +320,135 @@ func (s *Server) serveStreams(st transport.ServerTransport) {
 	wg.Wait()
 }
 
+var _ http.Handler = (*Server)(nil)
+
+// serveUsingHandler is the implementation of Serve(net.Listener) when
+// TestingUseHandlerImpl has been configured. This lets the end2end
+// tests exercise the ServeHTTP method as one of the environment types.
+func (s *Server) serveUsingHandler(lis net.Listener) error {
+	hs := &http.Server{
+		Handler:   s,
+		ConnState: s.onConnStateChange,
+	}
+	if err := http2.ConfigureServer(hs, &http2.Server{
+		MaxConcurrentStreams: s.opts.maxConcurrentStreams,
+	}); err != nil {
+		return err
+	}
+	hlis := &handlerListener{
+		s:        s,
+		Listener: lis,
+		acceptc:  make(chan interface{}, 1),
+		closedc:  make(chan struct{}),
+	}
+	go hlis.acceptLoop()
+	return hs.Serve(hlis)
+}
+
+// onConnStateChange is the net/http.Server.ConnState state change
+// hook used by the http.Handler-based transport, to track which
+// inbound TCP or TLS connections are live. Note that these are not
+// ServerTransports.  Each received HTTP request (each ServeHTTP call)
+// is a ServerTransport for exactly 1 stream. This on the other hand
+// tracks the underlying connections.
+func (s *Server) onConnStateChange(c net.Conn, state http.ConnState) {
+	if state != http.StateNew && state != http.StateClosed {
+		// Ignore transitions between idle and active.
+		return
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.hconns == nil {
+		s.hconns = make(map[net.Conn]bool)
+	}
+	if state == http.StateNew {
+		s.hconns[c] = true
+	} else {
+		delete(s.hconns, c)
+	}
+}
+
+type handlerListener struct {
+	s            *Server
+	net.Listener                  // embedded for Addr
+	acceptc      chan interface{} // of conn or error
+
+	closedc   chan struct{} // closed on close
+	closeOnce sync.Once
+	closeErr  error
+}
+
+func (hl *handlerListener) Close() error {
+	hl.closeOnce.Do(hl.closeOnceFunc)
+	return hl.closeErr
+}
+
+func (hl *handlerListener) closeOnceFunc() {
+	hl.closeErr = hl.Listener.Close()
+	close(hl.closedc)
+}
+
+func (hl *handlerListener) Accept() (net.Conn, error) {
+	select {
+	case v := <-hl.acceptc:
+		if c, ok := v.(net.Conn); ok {
+			return c, nil
+		}
+		return nil, v.(error)
+	case <-hl.closedc:
+		return nil, errors.New("listener closed")
+	}
+}
+
+// acceptLoop runs in its own goroutine and accepts conns and sets up
+// TLS, feeding successful connections to Accept.
+func (hl *handlerListener) acceptLoop() {
+	for {
+		rawConn, err := hl.Listener.Accept()
+		if err != nil {
+			select {
+			case hl.acceptc <- err:
```

Comment:
Since hl.authenticateConn(rawConn) is running in a separate goroutine, this could run into out-of-order issue:
at iteration N: a valid rawConn is generated and h1.authenticateConn is scheduled;
at iteration N+1: an error is returned by hl.Listener.Accept
But line#410 and line#432 could be executed in any order. Assume line#410 in the iteration N+1 gets executed first, then the "conn" in line#432 (iteration N) could be leaked -- the error got read and line#432 is executed (conn is written into hl.acceptc). But nobody will read from hl.acceptc again (http accept loop exits due to the error).